guibot/target.py

```diff
@@ -512,35 +512,66 @@ def load(self, steps_filename, **kwargs):
-                    raise IOError("Invalid chain step line '%s'" % dataconfig[0])
-                data, config = re.split(r'\t+', step.rstrip('\t\n'))
-
-                super(Chain, self).load(config)
-                self.use_own_settings = False
-
-                step_backend = self.match_settings.params["find"]["backend"]
-                if step_backend in ["autopy", "contour", "template", "feature", "tempfeat"]:
-                    data_and_config = Image(data, match_settings=self.match_settings)
-                elif step_backend in ["cascade", "deep"]:
-                    data_and_config = Pattern(data, match_settings=self.match_settings)
-                elif step_backend == "text":
-                    data_and_config = Text(data, match_settings=self.match_settings)
-                else:
-                    # in particular, we cannot have a chain within the chain since it is not useful
-                    raise UnsupportedBackendError("No target step type for '%s' backend" % step_backend)
-
-                self._steps.append(data_and_config)
+            lines = f.readlines()
+
+        while lines:
+            step = lines.pop(0)
+            dataconfig = re.split(r'\t+', step.rstrip('\t\n'))
+
+            # read a nested steps file and append to this chain
+            if dataconfig[0].endswith(".steps"):
+                nested_steps_filename = self.resolve_stepsfile(dataconfig[0])
+                # avoid infinite loops
+                if not nested_steps_filename in stepsfiles_seen:
+                    stepsfiles_seen.append(nested_steps_filename)
+                    with open(nested_steps_filename) as f:
+                        lines = f.readlines() + lines
+                continue
+
+            if len(dataconfig) != 2:
+                raise IOError("Invalid chain step line '%s'" % dataconfig[0])
+
+            data, config = dataconfig
+            super(Chain, self).load(config)
+            self.use_own_settings = False
+
+            step_backend = self.match_settings.params["find"]["backend"]
+            if step_backend in ["autopy", "contour", "template", "feature", "tempfeat"]:
+                data_and_config = Image(data, match_settings=self.match_settings)
+            elif step_backend in ["cascade", "deep"]:
+                data_and_config = Pattern(data, match_settings=self.match_settings)
+            elif step_backend == "text":
+                data_and_config = Text(data, match_settings=self.match_settings)
+            else:
+                # in particular, we cannot have a chain within the chain since it is not useful
+                raise UnsupportedBackendError("No target step type for '%s' backend" % step_backend)
+
+            self._steps.append(data_and_config)
 
         # now define own match configuration
         super(Chain, self).load(steps_filename)
 
+    def resolve_stepsfile(self, filename):
+        """
+        Try to find a valid steps file from a given file name.
```

Comment:
I am curious why the chains need this method and other targets don't. Is it just an artifact from refactoring the `load` method?

Reply:
Because other targets load files only once, while the chain with nested stepsfile support can load files multiple times.

Reply:
Could there be any situation where other targets could be loaded multiple times? You mean here it loads multiple times because of the nested step loading?

Reply:
> You mean here it loads multiple times because of the nested step loading?

Yes, that's the reason. This scenario is not valid for other targets

Reply:
In this case the `resolve_stepsfile` function could be internal to the `load` function or does the method really have other more general usage beyond the scope of the load method?

Reply:
I'd put it as a private method instead. Usually nested functions are useful when the inner is a closure or when we want to enforce encapsulation. But in this case it's not a closure and we're already inside a class, so the extra encapsulation brings no gains. The added indentation also decreases readability IMHO.

Reply:
What do you mean by the inner function being a closure? The usefulness of such nesting is that we can hide the function even further and restrict it only to its scope of application (the `load` method), essentially preserving the API the same. I guess choosing a private method instead of a more nested approach (a good feature of python providing even more information hiding if necessary) is a matter of taste but I am wondering about the real advantages of a private method as in the current situation is looks more like leaving internal artifacts on the same level with external methods.

Reply:
Funny, I was unfamiliar with the `closure` term. Regardless, the most important thing that we are both referring here is the scope of the function. By nesting it, we reduce its scope and keep the API simpler. But I leave this up to you since both methods work and have been used in the project.

Reply:
> The usefulness of such nesting is that we can hide the function even further and restrict it only to its scope of application (the load method)

Yep, this is "enforcing encapsulation". What I mean is that we are already hidden inside a class, so the scope is already limited and hiding the implementation of a function from methods of the same class doesn't bring much benefit. Also the API would remain the same as no extra public methods would be added - only a new private method, and these are not part of the API.

> I am wondering about the real advantages of a private method

It doesn't increase the indentation while keeping encapsulation pretty similar to what we already have. Deep nesting can [make bugs harder to find](https://testing.googleblog.com/2017/06/code-health-reduce-nesting-reduce.html) and is usually [reduces readability](https://en.wikibooks.org/wiki/Computer_Programming/Coding_Style/Minimize_nesting). Google [Python style guide](https://github.com/google/styleguide/blob/gh-pages/pyguide.md#26-nestedlocalinner-classes-and-functions) tells users to
> avoid nested functions or classes except when closing over a local value. Do not nest a function just to hide it from users of a module. Instead, prefix its name with an _ at the module level so that it can still be accessed by tests

However, the same style guide also cites
> Allows definition of utility classes and functions that are only used inside of a very limited scope.

as one of the advantages of nested functions. So I guess it's indeed a matter of personal preference. Having a background in C# (no nested functions until version 7) and Javascript (heavy use of closures), I guess I started to prefer it this way without noticing.

Reply:
I agree with the point about harder debugging and unit testing, I guess it is a valid one. OTOH even such guides remain opinions unless they offer more arguments like this. For instance, imagine a class with the other extreme of having many many methods where the time you will spend to find the ones that matter will increase each time you have to look up the class' API. In this case the structure becomes too linear, placing every detail on the same level of a hierarchy which could be richer in depth (not limited by the programming language). We thus end up losing encapsulation and showing more to the outside than we have to.

But let's not forget that python like any other technology is a product of human evolution of design and thus there are many partially developed concepts that could keep moving in one direction or the other.

Reply:
Just to comment more on here:

> class with the other extreme of having many many methods where the time you will spend to find the ones that matter will increase each time you have to look up the class' API

Indeed, in this case extending the class would make browsing harder. But I see this as hint of an even worse problem: the class is most likely doing too much and should be split into smaller classes ([SRP principle](https://medium.com/@severinperez/writing-flexible-code-with-the-single-responsibility-principle-b71c4f3f883f)). But I usually take these recommendations with a grain of salt; I've seen them being interpreted too literally resulting in messy commits introducing dozens of 5-line classes.

Anyway, I guess this is EOPR ;) Thanks for merging!

Reply:
Hehe, I knew this would come and expected to hear it :joy: But I am glad you pointed at the downside of this approach too. When I separate classes I prefer to follow a conceptual criterion rather than a plain "too many lines" criterion. But this is a matter of how we interpret OOP and the orthodox conceptual class representations are less modern for sure.

EOPR indeed! Thanks for the patience!